geowave/src/main/scala/geotrellis/spark/io/geowave/GeowaveLayerWriter.scala

```diff
@@ -0,0 +1,253 @@
+package geotrellis.spark.io.geowave
+
+import geotrellis.geotools._
+import geotrellis.proj4.LatLng
+import geotrellis.raster._
+import geotrellis.spark._
+import geotrellis.spark.io._
+import geotrellis.spark.io.accumulo._
+import geotrellis.spark.io.avro._
+import geotrellis.spark.io.index.KeyIndex
+import geotrellis.util._
+import geotrellis.vector.Extent
+
+import com.typesafe.scalalogging.slf4j._
+import mil.nga.giat.geowave.adapter.raster.adapter.merge.RasterTileRowTransform
+import mil.nga.giat.geowave.adapter.raster.adapter.RasterDataAdapter
+import mil.nga.giat.geowave.core.geotime.ingest._
+import mil.nga.giat.geowave.core.store._
+import mil.nga.giat.geowave.core.store.adapter.statistics.StatsCompositionTool
+import mil.nga.giat.geowave.core.store.data.VisibilityWriter
+import mil.nga.giat.geowave.core.store.operations.remote.options.DataStorePluginOptions
+import mil.nga.giat.geowave.core.store.util.DataStoreUtils
+import mil.nga.giat.geowave.datastore.accumulo._
+import mil.nga.giat.geowave.datastore.accumulo.metadata._
+import mil.nga.giat.geowave.datastore.accumulo.operations.config.AccumuloOptions
+import mil.nga.giat.geowave.datastore.accumulo.util._
+import mil.nga.giat.geowave.datastore.accumulo.util.AccumuloUtils
+import org.apache.accumulo.core.client.security.tokens.PasswordToken
+import org.apache.accumulo.core.data.{ Key, Value }
+import org.apache.spark.rdd.RDD
+import org.apache.spark.SparkContext
+import org.geotools.coverage.grid.GridCoverage2D
+import org.geotools.coverage.processing.CoverageProcessor
+import org.geotools.factory.{ GeoTools, Hints }
+import org.opengis.coverage.grid.GridCoverage
+import org.opengis.parameter.ParameterValueGroup
+
+import java.io.{ DataInputStream, DataOutputStream, File }
+import java.util.UUID
+import javax.imageio.ImageIO
+import javax.media.jai.{ ImageLayout, JAI }
+import scala.collection.JavaConversions._
+import scala.collection.mutable.ListBuffer
+import scala.reflect._
+
+import resource._
+import spray.json._
+
+
+object GeowaveLayerWriter extends LazyLogging {
+
+  def write[
+    K: ClassTag,
+    V: ClassTag,
+    M: JsonFormat: GetComponent[?, Bounds[K]]
+  ](
+    coverageName: String,
+    rdd: RDD[(K, V)] with Metadata[M],
+    as: GeowaveAttributeStore,
+    accumuloWriter: AccumuloWriteStrategy
+  ): Unit = {
+    val metadata = implicitly[ClassTag[K]].toString match {
+      case "geotrellis.spark.SpatialKey" => rdd.metadata.asInstanceOf[TileLayerMetadata[SpatialKey]]
+      case t: String => throw new Exception(s"Unsupported Key Type: $t")
+    }
```

Comment:
Again, we should encode this in the type system instead of making this a runtime error, even though that means breaking off from the LayerReader/Writer trait structure.